engine/grid.py

```diff
@@ -226,70 +228,69 @@ def activate_circle(self, circle_center_row, circle_center_col, circle_radius):
 
+
+ 
+ 
     # A function to check whether point P(x, y)
     # lies inside the triangle formed by
     # A(x1, y1), B(x2, y2) and C(x3, y3)
-
-    def is_inside_triangle(self, x1, y1, x2, y2, x3, y3, x, y):
+    def isInsideTriangle(self, x1, y1, x2, y2, x3, y3, x, y):
         def area(x1, y1, x2, y2, x3, y3):
-
+    
             return abs((x1 * (y2 - y3) + x2 * (y3 - y1)
                         + x3 * (y1 - y2)) / 2.0)
-
+        
         # Calculate area of triangle ABC
         A = area(x1, y1, x2, y2, x3, y3)
-
+    
         # Calculate area of triangle PBC
         A1 = area(x, y, x2, y2, x3, y3)
-
+        
         # Calculate area of triangle PAC
         A2 = area(x1, y1, x, y, x3, y3)
-
+        
         # Calculate area of triangle PAB
         A3 = area(x1, y1, x2, y2, x, y)
-
+        
         # Check if sum of A1, A2 and A3
         # is same as A
-        if (A == A1 + A2 + A3):
+        if(A == A1 + A2 + A3):
             return True
         else:
             return False
-
-    # Activate triangle based on three points
-
-    def activate_traingle(self, x1, y1, x2, y2, x3, y3):
-
+    ##Activate triangle based on three points
+    def activate_triangle(self,x1,y1,x2,y2,x3,y3):
         """
         Activates all the nodes in a traingle.
         """
         rows = self.nodes.shape[0]
         cols = self.nodes.shape[0]
         for x in range(rows):
             for y in range(cols):
-                if self.isInsideTriangle(x1, y1, x2, y2, x3, y3, x, y):
-                    self.nodes[x, y].is_active = True
-
-    # Checks if node is on border of activated nodes
-    def is_on_border(self, row, col, row_limit, col_limit):
-        min_col = max(0, col-1)
-        min_row = max(0, row-1)
-        max_col = min(col_limit, col+1)
-        max_row = min(row_limit, row+1)
-
-        # If this node is on the very edge of the grid, it is automatically a border node
-        if min_col == 0 or min_row == 0 or max_col == col_limit or max_row == row_limit:
-            return True
-
-        # If the node has a neighboring node that is inactive, it is a border node
-        for col in range(min_col, max_col+1):
-            for row in range(min_row, max_row+1):
-                if not self.nodes[row][col].is_active:
-                    return True
-        return False
+                if self.isInsideTriangle(x1,y1,x2,y2,x3,y3,x,y):
+                    self.nodes[x,y].is_active = True
+
+    ##Checks if node is on border of activated nodes
+    def is_on_border(self,row, col, row_limit, col_limit):
+            min_col = max(0, col-1)
+            min_row = max(0, row-1)
+            max_col = min(col_limit, col+1)
+            max_row = min(row_limit, row+1)
+
+            # If this node is on the very edge of the grid, it is automatically a border node
+            if min_col == 0 or min_row == 0 or max_col == col_limit or max_row == row_limit:
+                return True
```

Comment:
Thinking about this bc we have double borders if we try to use almost all the grid points (start row/col = 0, end row/col = rows/cols -1)
I tried
```
        if row == 0 or col == 0 or col == self.get_num_cols or row == self.get_num_rows:
            return True
```
and it fixed the double borders but there is a problem w the circle border (see pics below)

I think work on this can be in a separate PR, unless you think this could affect the current impl for lawnmower!

Reply:
Before
<img width="708" alt="Screen Shot 2022-11-14 at 2 53 14 AM" src="https://user-images.githubusercontent.com/60713782/201605085-4624a8ed-9b43-4d9a-a860-453b3c09bc43.png">
<img width="708" alt="Screen Shot 2022-11-14 at 2 53 30 AM" src="https://user-images.githubusercontent.com/60713782/201605089-6c7a18a5-f139-4c79-8e88-4659bf6fd558.png">

After
<img width="708" alt="Screen Shot 2022-11-14 at 2 52 25 AM" src="https://user-images.githubusercontent.com/60713782/201605165-b6b129e3-022c-497f-83af-8173cd41ed6c.png">
<img width="708" alt="Screen Shot 2022-11-14 at 2 53 43 AM" src="https://user-images.githubusercontent.com/60713782/201605167-1f41e30d-6c86-4684-9a3e-79097de36ac6.png">